File: 프로그래머스/unrated/181918. 배열 만들기 4/배열 만들기 4.java
import java.util.*;

public class Solution {
    public int[] solution(int[] arr) {
        Stack<Integer> stack = new Stack<>();
        int i = 0;
        while (i < arr.length){
            if (stack.isEmpty()){
                stack.push(arr[i]);
                i++;
            } else if (stack.peek() < arr[i]) {
                stack.push(arr[i]);
                i++;
            } else {
                stack.pop();
            }
        }
        return stack.stream().mapToInt(Integer::intValue).toArray();
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        int[] arr = {1, 4, 2, 5, 3};
        int[] answer = sol.solution(arr);
        System.out.println(Arrays.toString(answer));
    }
}

File: JavaScript/프로그래머스/0/181918. 배열 만들기 4/배열 만들기 4.js
function solution (arr) {
    let answer = [];
    let i = 0;

    while (i < arr.length) {
        if (answer.length === 0) {
            answer.push(arr[i]);
            i++;
        } else if (answer[answer.length - 1] < arr[i]) {
            answer.push(arr[i]);
            i++;
        } else {
            answer.pop();
        }
    }

    return answer;
};